doc/tally-recovery.md

```diff
@@ -0,0 +1,95 @@
+order of fragments is used to decide which choice applies when multiple ballots
+for the same proposal have been received from a single account.
+
+The format of the log entries is bincode serialization of a Rust structure
+containing the unsigned integer timestamp in seconds since the Unix epoch,
+and the fragment serialized in the binary blockchain format: 
+
+```rust
+pub struct FragmentLogEntry {
+    /// The time this fragment was registered and accepted by the pool
+    pub time: SecondsSinceUnixEpoch,
+    /// Fragment body serialized as byte array
+    pub fragment: RawFragment,
+}
+```
+
+## Initial state: voter accounts in block0
+
+To determine which accounts can legitimately vote, the tool parses the
+genesis block of the Catalyst blockchain. It is assumed that the registered
+voters' account balances are created in the initial fund distribution in the
+genesis block. The initial ledger state is recreated from the genesis block
+just like in the blockchain node (with some technical twists: a
+mirror set of accounts is created because we need some private keys to
+supply to the library code performing ledger state transitions).
+
+## Replay of transactions
+
+The tally recovery command parses the persistent fragment logs and applies
+the fragments in order they were received, to replay changes in the ledger state
+of a blockchain node that would process transactions to the same effect, if
+the account spending counters used to sign transactions spending form the same
+account incremented in order of submission by the client.
+
+Transaction validity checks are performed just like in the blockchain node
+validating transactions for a block, with the exception that the witness
+signature check for account inputs is performed repeatedly with multiple
+candidate values of the spending counter until a matching signature is found
+or the search range is exhausted, searching with increasing difference
+above and below the next expected counter value for the account
+as per the normal transaction validation rule. This ensures that transactions
+submitted by a participant in control of the account private key, but
+with out-of-order values of the spending counter, are processed as valid.
+
+Potential replay attacks are prevented by keeping a record of all fragment
+hashes encountered and rejecting any duplicate fragments. It is assumed that
+the likelihood of a client submitting two legitimate ballots with the same
+proposal and choice and the same (incorrectly used) spending counter value
+is low enough to ignore such occurrences.
+
+To simplify the logic, the recovery tool only processes fragments of
+these kinds:
+
+- Ballot (VoteCast) transactions with one account input and no outputs.
+  Private votes are not yet supported.
+- Plain value transfer transactions with one account input and one account
+  output. The recovery tool updates the voting power with the results of
+  the transfer and adds the output accounts.
```

Comment:
The tally tool also processes `VoteTally` transactions. To be precise, those are a bit different from other transactions in the sense that they are replayed without the use of mirror accounts.